### ISSUE TITLE
(feat req) Add support for dependency rules

### ISSUE THREAD

**Issue by hulkish** [renovatebot/renovate#1557]
## Summary
I would like renovate to make pr's based on rules that I configure which determine *how* particular dependencies are defined.

### 1. Banning/Removing select dependencies
When I want to remove/prevent use of a particular dependency **everywhere**. For example, if we wanted to remove `lodash` everywhere:
```json5
  "dependencyRules": [
    {
      "patterns": ["^lodash$"],
      "resolvers": [
        "remove",
        // or, only remove from a particular dependency category:
        // ["remove", ["dependencies"]]
      ]
    }
  ]
```

### 2. Correct mistakes
Renovate could help with bringing visibility common mistakes made when defining dependencies.

Example: When somebody defines `babel-core` in `dependencies` - it's almost always a **mistake**. Here's how your config might look in order for renovate to rectify the problem:
```json5
  "dependencyRules": [
    {
      "patterns": [
        "^babel"
       ],
      "resolvers": [
        ["moveTo", ["devDependencies"]]
      ]
    },
    {
      "patterns": [
        "babel-runtime",
        "regenerator-runtime"
      ],
      "resolvers": [
        ["moveTo", ["devDependencies", "peerDependencies"]],
        ["forceRange", {
           "dependencies": "^",// default
           "devDependencies": "",.// default
           "peerDependencies": ">="
        }]
      ]
  ]
```

### 3. Assign dependency types
When I want particular dependencies to only be defined in `peerDependencies`:
```json5
  "packageRules": [
    {
      "packagePatterns": ["^@myscope"],
      "groupName": "myscope packages",
      "dependencyRules": [
        {
          "patterns": ["^react"],
          "resolvers": [
            ["moveTo", ["devDependencies", "peerDependencies"]]
          ]
        }
      ]
    }
  ]
```

### 4. Assign ranges to select dependencies
When I want particular dependencies to be defined with a specific version & range:
```json5
  "packageRules": [
    {
      "packagePatterns": ["^@myscope"],
      "groupName": "myscope packages",
      "dependencyRules": [
        {
          "patterns": ["^react"],
          "resolvers": [
            ["forceRange", {
               "dependencies": "^",// default
               "devDependencies": "",.// default
               "peerDependencies": ">="
            }],
            ["forceVersion", "15.0.0"]
          ]
        }
      ]
    }
  ]
```

### 5. Force versions for select dependencies
When I want to enforce a certain version of a dependency:
```json5
"dependencyRules": [
    {
      "patterns": [
        "!babel-runtime",
        "!regenerator-runtime"
      ],
      "resolvers": [
        ["moveTo", ["devDependencies", "peerDependencies"]],
        ["forceRange", {
           "dependencies": "^",// default
           "devDependencies": "",.// default
           "peerDependencies": ">="
        }]
      ]
  ]
```

## Config API

- This feature would make `dependencyRules` available at every tier of your config. This includes global, repo, devDependencies, peerDependencies, packageRules, etc.
- A dependency rule is an object consisting of 2 possible fields: `patterns` & `resolvers`.
  - `patterns` 
    - Patterns to match against the `name` field in each `package.json`.
  - `resolvers` 
    - Strategy applied to the matched rule. Potential resolver types could be:
      - `add`: Add a dependency, per dependency type.
      - `remove`: Remove/Ban per dependency type.
      - `moveTo`: Constrain allowed dependencies types. (Maybe this one can be replaced with `add` & `remove`?)
      - `forceRange`: What range should be used, per dependency type
      - `forceVersion`: What version should be used, per dependency type.

## Notes
I realize this idea has many other possible variations. I have described the most prominent aspects of this feature that apply to what I'm looking for. 

But, I expect this idea could be talked through a bit before going forward. So, I'd like to get some buy-in here and see if there is interest - before contributing to this myself.

Some things to consider, still:
- How does this impact existing features?
- Would it make sense to maybe group this behavior with "lockfile maintenance", or more of a general "maintenance" category?

After reading through a couple issues/feature requests on this repo. I feel like the proposed solution above would assist with a couple ive come across:
- https://github.com/renovateapp/renovate/issues/1555
- https://github.com/renovateapp/renovate/issues/1503

**Comment by roastlechon** [comment]
I'm interested in this too

**Comment by amalgamet** [comment]
+1

**Comment by rarkins** [comment]
> 1. Banning/Removing select dependencies

I'm not sure this is necessarily a good fit for Renovate, but willing to consider it. Usually something like `lodash` is in projects for a reason - like any other dependency should be - so simply removing it will break the project. So if we remove it while it's still actually used, we've done nothing but raise a broken PR.

I would suggest something like #1503 as an alternative, unless there's another reason for this feature that I'm missing.

**Comment by rarkins** [comment]
> 2. Correct mistakes

My instinct is to prefer this as part of someone ele's linting package, rather than Renovate. It's not a completely bad fit though, however I think raising PRs for things which might just be warnings could be annoying for some if these would be better placed as tests.

**Comment by rarkins** [comment]
> 3. Assign dependency types

This seems almost the same as (2) to me. The example looks a little strange because you have a "^@myscope" pattern as well as a "^react" pattern within the same rule. Might help me if you can edit the above to explain what this actual example is meant to mean? e.g. it is specific to your private scope? Or is it so that any ^react package is left out of `dependencies` and only in `devDependencies` + `peerDependencies`?

**Comment by rarkins** [comment]
> 4. Assign ranges to select dependencies

The purpose of this is to enforce the same version of react across all of your repositories?

I think it might be able to use the same solution as #1555

**Comment by rarkins** [comment]
> 5. Force versions for select dependencies

Was this meant to use the example from (4) ?

**Comment by rarkins** [comment]
Thanks @hulkish for raising this. If you can expand the above examples to help me understand your use cases better (e.g. I'm guessing most of them are about you wanting to enforce policies and exact versions across multiple repositories).

@destroyerofbuilds I know you also work a lot of enforcing consistency across repositories, have you had similar requirements?

**Comment by hulkish** [comment]
@rarkins Not only version, but the range and how it's defined is also important. For example... Per this example, react would be preferred to be defined as a peer dependency with range/version of `>=15.0.0`

**Comment by hulkish** [comment]
I'll follow up later today with some more details on your questions above.

**Comment by hulkish** [comment]
>> 1. Banning/Removing select dependencies

> I'm not sure this is necessarily a good fit for Renovate, but willing to consider it. Usually something like lodash is in projects for a reason - like any other dependency should be - so simply removing it will break the project. So if we remove it while it's still actually used, we've done nothing but raise a broken PR.

I view these pr's in the same way I view a pr for a major version bump on a dependency:
- Will only actually be merged after careful consideration & probably some manual intervention.
- The attention to detail required to do this manually across several packages (100+) quickly starts to stack up as something you can't realistically maintain on a regular basis.
- I agree this is a unique & probably not common issue. Most of these use cases are non-issues in the context of managing 1-10 packages.
- Keep in mind that all of my uses cases in this issue are focused on managing multiple repos/packages. If you've never encountered this, then you are probably thinking "just handle it manually, no big deal". Well that is a failure to understand how much work this actually is.

> I would suggest something like #1503 as an alternative, unless there's another reason for this feature that I'm missing.

While that behavior is desired - it doesn't quite satisfy the need. I also need a guarantee of *how* the dependency is defined (dependencies, peerDependencies, etc). 

>> 2. Correct mistakes

> My instinct is to prefer this as part of someone ele's linting package, rather than Renovate. It's not a completely bad fit though, however I think raising PRs for things which might just be warnings could be annoying for some if these would be better placed as tests.

This was actually my original approach for this. However, this still leads to manually correcting these manually. Linting these problems - while useful, does nothing to actually help you solve this problem. 

The reason I feel it's a good fit for renovate for the the same reasons `pinVersions` behavior is helpful - because trying to handle this manually is not practical.

>> 4. Assign dependency types

> This seems almost the same as (2) to me. The example looks a little strange because you have a "^@Myscope" pattern as well as a "^react" pattern within the same rule. Might help me if you can edit the above to explain what this actual example is meant to mean? e.g. it is specific to your private scope? Or is it so that any ^react package is left out of dependencies and only in devDependencies + peerDependencies?

I think maybe it's a little confusing because it looks similar to other use cases. This is just to illustrate the behavior of the dependency rule resolver. So, yes to *"Or is it so that any ^react package is left out of dependencies and only in devDependencies + peerDependencies?"*.

Basically the examples above which use `packageRules` & `dependencyRules` together intend to grant you the ability to isolate these dependency rules to only a few of your own packages.

**Comment by rarkins** [comment]
OK, I'll keep thinking about this further and how it might fit in most elegantly. A term that has come to mind for this type of stuff is "enforce"/"enforcement". Or maybe "policy enforcement".

I'm still a bit uncertain about the lodash example. Basically you're saying "remove this from every repository even if they're still using it". For repos not using it, it will be a PR that should pass tests. For those using it, it will be a broken PR. But that's what you want?

**Comment by hulkish** [comment]
@rarkins To be fair, the lodash is probably my poorest example of where this is useful - I added that use case to better illustrate the goal.

**Comment by hulkish** [comment]
@rarkins Also, I wanted to expand specifically on the use case involving the need to ensure certain deps are defined inside `peerDependencies`. Consider the following:

You have an app, which collectively involves 2 repositories:
- `foo-app-repo`
  - Contains a single `package.json`. including some public and private dependencies. 
  - All private packages use `@foo` scope.
- `foo-libs-monorepo`
  - Contains over 100 `@foo/` packages. 
  - Most of these packages define react, react-dom (for ex) inside `dependencies`.
  - Some of these packages were created at different times. 
    - Many define react without a version range or `~`, some use `^`.

Per above, when you `npm`/`yarn` install inside `foo-app-repo` - you will find that there is a world of incorrect react versions being used. Many nested `node_modules/react` folders which are (incorrectly) using different react versions.

Some signs that indicate "should this be a peer dependency":
- Your app or top-level package defines the same dep inside `dependencies` or `devDependencies` (if it's a dev package, for example)
- Many other dependencies already define the particular dep as a peer.

So, as a way to "auto determine" what should be a peer, I wonder if the following would help:
1. Do a lookup in your app/top level `package.json` for what is defined inside `dependencies` & `devDependencies`.
2. Any intersecting deps in your `@foo` packages should be defined as peer.

**Comment by destroyerofbuilds** [comment]
> @destroyerofbuilds I know you also work a lot of enforcing consistency across repositories, have you had similar requirements?

We have similar needs.

As outlined in #1555, we need to enforce a AngularJS usage policy across our organization.

We also expect developers to remove unused dependencies following a refactor.

Lastly, there are mono-repository teams managing many small libraries, and I would bet they would appreciate a way to enforce consistent versioning across their `package.json` files.

However, as I [outlined here](https://github.com/renovateapp/renovate/issues/1503#issuecomment-368619287) I don't see an advantage of managing those policies and best practices through the Renovate service.

I would prefer a linter, such as [`eslint-plugin-importer`](https://www.npmjs.com/package/eslint-plugin-import), enforce these policies.

On a related note, I would like to see the ability to add addition _fixable_ lint rules to an ESLint configuration package (Which would be a major breaking change since new enforcement rules will break linting), and see those get automatically fixed by Renovate as part of an upgrade - #838 

> Linting these problems - while useful, does nothing to actually help you solve this problem.

@hulkish many plugins for ESLint have [autofixers](https://eslint.org/docs/user-guide/command-line-interface#--fix).

Looks like there's a discussion about [adding fixers](https://github.com/benmosher/eslint-plugin-import/issues/547) for `eslint-plugin-import`. (Though `eslint-plugin-import` itself does not actually address any of the scenarios mentioned above)
